### PR TITLE
Proposing identifiers.

### PR DESCRIPTION
**Pull request by svenboulanger**
Proposal for allowing more freedom involving identifiers.

### CODE REVIEW

File: SpiceSharp/Simulations/Variables/VariableSet.cs [hunk 1]
```diff
@@ -56,8 +56,29 @@ public VariableSet()
...
+
+            // Unlock
+            _locked = false;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="VariableSet"/> class.
+        /// </summary>
+        /// <param name="comparer">The comparer for identifiers.</param>
+        public VariableSet(IEqualityComparer<Identifier> comparer)
```
[the remark this answers is not in the document]
Comment (marcin-golebiowski) [reply SpiceSharp/Simulations/Variables/VariableSet.cs]: @svenboulanger  And I think it should have ...

**Comment by svenboulanger** [comment]
I am probably going to avoid specifying these things in the constructor of `Simulation`. Instead, I'll add an *optional* configuration that allows you to specify the equality comparer.

We can do that because the objects in question are only built in the `Setup()` method, at that point we can read the configuration and build the lists.

File: SpiceSharp/Circuits/Circuit.cs [hunk 2]
```diff
@@ -41,6 +41,16 @@ public class Circuit
...
         public Circuit()
         {
+            Entities = new EntityCollection();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Circuit"/> class.
+        /// </summary>
+        /// <param name="comparer">The comparer for identifiers.</param>
+        public Circuit(IEqualityComparer<Identifier> comparer)
```
[the remark this answers is not in the document]
Comment (svenboulanger) [reply SpiceSharp/Circuits/Circuit.cs]: I have done this by adding the `CollectionsConfiguration` class which can be added in the `Configuration` of the simulation.

**Comment by svenboulanger** [comment]
There are also other issues. Some behaviors create a new variable/node where they use `new SubIdentifier(..., "branch")`. In order for the second parameter to use a custom identifier, we need some kind of factory.

**Comment by svenboulanger** [comment]
We can take this idea further and also add a similar construction for other implicit conversion types ( eg. `int` and/or `long`). What do you think? @marcin-golebiowski

**Comment by marcin-golebiowski** [comment]
@svenboulanger
I think factory idea with static method will not allow me to introduce different case-sensitivity settings for nodes, entities, parameters, function names. Maybe that feature is not really needed but will make the parser more powerful.

**Comment by marcin-golebiowski** [comment]
@svenboulanger  Maybe we could combine two ideas.
I could introduce "Default setting" for everything else that will be controlled with that static factory and everything else with proper comparers. I'm not sure.

**Comment by svenboulanger** [comment]
@marcin-golebiowski Maybe we are over-designing identifiers? We can also just move to plain strings. Simple and easy.

**Comment by marcin-golebiowski** [comment]
@svenboulanger  I like the idea 👍 I vote for no implicit conversions. However the comparers should stay.

**Comment by svenboulanger** [comment]
@marcin-golebiowski I removed `Identifier` and replaced everything with normal strings.
One note though:
Some behaviors create an extra variable that has to be named. I made this somewhat configurable by adding the static `Combine` method, which just concatenates strings with an optional separator.

**Comment by marcin-golebiowski** [comment]
Works like charm :) 
However some other changes are required (see pull request #43)

**Comment by svenboulanger** [comment]
I'll go ahead and merge this into the development branch.

**Comment by marcin-golebiowski** [comment]
@svenboulanger  Feel free to do it 👍

**Pull request merged**